File: Python3/201 - 400/384 Shuffle an Array.py
# 412ms, 99.50%
class Solution:
    def __init__(self, nums):
        """
        :type nums: List[int]
        """
        self.origin = nums
        self.shuffled = nums[:]

    def reset(self):
        """
        Resets the array to its original configuration and return it.
        :rtype: List[int]
        """
        return self.origin

    def shuffle(self):
        """
        Returns a random shuffling of the array.
        :rtype: List[int]
        """
        # use random.SystemRandom and costs 1368ms
        random.shuffle(self.shuffled)
        return self.shuffled


# Your Solution object will be instantiated and called as such:
# obj = Solution(nums)
# param_1 = obj.reset()
# param_2 = obj.shuffle()

File: 201 - 400/384 Shuffle an Array.cc
// 215ms, 99.85%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false); // toggles off the synchronization
    cin.tie(nullptr);                 // ties cin with nullptr, not cout
    return 0;
}();

class Solution {
private:
    vector<int> origin;
    vector<int> shuffled;

public:
    Solution(const vector<int> &nums) : origin(nums), shuffled(nums) {}

    /** Resets the array to its original configuration and return it. */
    vector<int> reset() {
        return origin;
    }

    /** Returns a random shuffling of the array. */
    vector<int> shuffle() {
        std::random_shuffle(shuffled.begin(), shuffled.end());
        return shuffled;
    }
};

/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * vector<int> param_1 = obj.reset();
 * vector<int> param_2 = obj.shuffle();
 */
